### PR TITLE
MueLu: Adding CM/RCM support to MueLu::Utilities_kokkos

### PR DESCRIPTION
**Pull request by csiefer2**
Wraps @brian-kelley's RCM code in Kokkos-Kernels

Tests: RHEL 6

### CODE REVIEW

**Comment by brian-kelley** [comment]
@csiefer2 @lucbv It works, but it's slow on GPU. I do know how I would write a much faster BFS but I need to finish the cluster GS rewrite first.

Also, the RCM object can do ``rcm()`` for reverse order, or ``cuthill_mckee()`` for forward. So you don't have to call ``rcm()`` and then reverse it in your own code.

**Comment by brian-kelley** [comment]
I'm really itching to do a fast BFS because it could also easily generalize to single-source shortest path, widest path, coarsening and possibly the sparse triangular solves in KK :)

File: packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp [hunk 1]
```diff
@@ -636,6 +652,70 @@ namespace MueLu {
...
+    return retval;
+  }
+  
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  Teuchos::RCP<Xpetra::Vector<LocalOrdinal,LocalOrdinal,GlobalOrdinal,Node> > CuthillMcKee(Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> &Op) {
+    using local_matrix_type = typename Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>::local_matrix_type;
+    using local_graph_type  = typename local_matrix_type::staticcrsgraph_type;
+    using device            = typename local_graph_type::device_type;
+
+    Teuchos::RCP<Xpetra::Vector<LocalOrdinal,LocalOrdinal,GlobalOrdinal,Node> > ordering = MueLu::ReverseCuthillMcKee(Op);
```
Comment (brian-kelley) [inline packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp:687]: Here, could do
```
    rcm_t rcm(localMatrix.numRows(), localMatrix.graph.row_map, localMatrix.graph.entries);
    lno_nnz_view_t cmOrder = rcm.cuthill_mckee(); 
```
and then don't need to reverse it.

Comment (csiefer2) [reply packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp:687]: Ooh!  I missed that!  Thanks for the pointer!

**Comment by csiefer2** [comment]
> 
> 
> @csiefer2 @lucbv It works, but it's slow on GPU. I do know how I would write a much faster BFS but I need to finish the cluster GS rewrite first.


Yeah, we don't really need performance at current.  But when we do, I'll be sure to let you know!

File: packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp [hunk 2]
```diff
@@ -103,8 +104,23 @@
 
 #include <MueLu_Utilities_kokkos_decl.hpp>
 
+#include <KokkosKernels_Handle.hpp>
+#include <KokkosSparse_partitioning_impl.hpp>
+
+
 namespace MueLu {
 
+  template<class ViewType, class execution_space = typename ViewType::execution_space>
```
Comment (brian-kelley) [inline packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp:113]: Oh I missed this. We could just delete this now, and also you're swapping ``i`` and ``N-2``, not ``N-1-i``.

Comment (csiefer2) [reply packages/muelu/src/Utils/MueLu_Utilities_kokkos_def.hpp:113]: Will delete on next PR (when I integrate this goop into the Pairwise aggregation.

**Comment by jhux2** [comment]
> I'm really itching to do a fast BFS because it could also easily generalize to single-source shortest path, widest path, coarsening and possibly the sparse triangular solves in KK :)

It could also be used in Zoltan2!

**Comment by kddevin** [comment]
@egboman @srajama1 @csiefer2 

Are you adding new ordering capability to @trilinos/muelu ?  
Zoltan2 has an ordering interface.  What are your needs?

**Comment by lucbv** [comment]
@kddevin 
Chris and I are working on an aggregation algorithms that is sensitive to local ordering.
Here we are just creating interfaces to call the cm and rcm from kokkos-kernels on `KokkosSparse::CrsMatrix`.
I think that for these local operations it is probably appropriate to call kk directly instead of Zoltan2?

**Pull request merged**